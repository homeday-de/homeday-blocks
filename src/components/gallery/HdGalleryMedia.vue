<template>
  <div class="gallery-media">
    <div class="gallery-media__object">
      <div
        :style="sizerStyles"
        class="gallery-media__object__sizer"/>
      <div
        v-if="item.image"
        ref="backgroundHolder"
        :style="{
          'background-image': `url('${item.image}')`,
        }"
        class="gallery-media__object__background"
      />
      <div
        v-if="hasThumbnail"
        :class="{
          'gallery-media__object__thumbnail': true,
          'isVisible': showThumbnail,
        }"
        :style="{
          'background-image': `url('${item.thumbnail}')`,
        }"
      />
    </div>
  </div>
</template>

<script>
import imagesLoaded from 'imagesloaded';

export default {
  name: 'HdGalleryMedia',
  props: {
    item: {
      type: Object,
      default: () => ({}),
    },
    aspectRatio: {
      type: Number,
      default: 16 / 9,
    },
  },
  data() {
    return {
      showThumbnail: true,
    };
  },
  computed: {
    hasThumbnail() {
      return typeof this.item.thumbnail === 'string' && this.item.thumbnail;
    },
    sizerStyles() {
      return {
        paddingTop: `${100 / this.aspectRatio}%`,
      };
    },
  },
  watch: {
    item() {
      this.$nextTick(() => {
        this.hideThumbnailOnBgLoad();
        this.showThumbnail = true;
      });
    },
  },
  mounted() {
    this.hideThumbnailOnBgLoad();
  },
  methods: {
    imageLoadListener() {
      this.showThumbnail = false;
    },
    hideThumbnailOnBgLoad() {
      imagesLoaded(this.$refs.backgroundHolder, { background: true }, () => {
        this.showThumbnail = false;
      });
    },
  },
};
</script>

<style lang="scss">
@import 'hd-blocks/styles/mixins.scss';

.gallery-media {
  $_root: &;

  &__object {
    position: relative;
    overflow: hidden;
    border-radius: 2px;

    &__background {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      background-color: $wild-sand;
      background-position: center;
      background-repeat: no-repeat;
      background-size: contain;
    }

    &__thumbnail {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      background-position: center;
      background-repeat: no-repeat;
      background-size: contain;
      opacity: 0;
      transition: opacity ($time-s * 2) ease-in-out;
      filter: blur(4px);

      &.isVisible {
        opacity: 1;
        transition: none;
      }
    }
  }
}
</style>
