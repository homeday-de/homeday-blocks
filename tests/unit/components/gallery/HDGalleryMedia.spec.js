import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdGalleryMedia from '@/components/gallery/HdGalleryMedia.vue';
import ITEMS from '@/stories/mocks/GALLERY_ITEMS';

const wrapperBuilder = wrapperFactoryBuilder(HdGalleryMedia, {
  props: {
    item: ITEMS[0],
  },
});

describe('HdGalleryMedia', () => {
  it('renders component', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
