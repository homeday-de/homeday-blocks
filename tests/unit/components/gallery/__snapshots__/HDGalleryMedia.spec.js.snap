// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdGalleryMedia renders component 1`] = `
<div class="gallery-media">
  <div class="gallery-media__object">
    <div class="gallery-media__object__sizer" style="padding-top: 56.25%;"></div>
    <picture class="gallery-media__object__picture">
      <source media="(max-width: 600px)" srcset="https://dummyimage.com/306x172?text=1+mw+600px">
      <source media="(max-width: 900px)" srcset="https://dummyimage.com/768x432?text=1+mw+900px">
      <source media="(max-width: 1200px)" srcset="https://dummyimage.com/768x432?text=1+mw+1200px"> <img src="https://dummyimage.com/1536x364?text=1" alt="Image 1"></picture>
    <div class="gallery-media__object__thumbnail isVisible" style="background-image: url(https://dummyimage.com/154x86?text=1+thumb);"></div>
  </div>
</div>
`;
